operator/src/main/java/oracle/kubernetes/operator/steps/ReadHealthStep.java

```diff
@@ -128,47 +242,62 @@ public NextAction apply(Packet packet) {
 
           String serviceUrl =
-              HttpClient.getServiceUrl(
+              getServiceUrl(
                   service,
                   pod,
                   serverConfig.getAdminProtocolChannelName(),
                   serverConfig.getListenPort());
           if (serviceUrl != null) {
-            Result result =
-                httpClient.executePostUrlOnServiceClusterIP(
-                    getRetrieveHealthSearchUrl(),
-                    serviceUrl,
-                    getRetrieveHealthSearchPayload(),
-                    false);
-
-            Pair<String, ServerHealth> pair = createServerHealthFromResult(result);
-
-            String state = pair.getLeft();
-            if (state != null && !state.isEmpty()) {
-              ConcurrentMap<String, String> serverStateMap =
-                  (ConcurrentMap<String, String>) packet.get(SERVER_STATE_MAP);
-              info.updateLastKnownServerStatus(serverName, state);
-              serverStateMap.put(serverName, state);
-            }
-
-            @SuppressWarnings("unchecked")
-            ConcurrentMap<String, ServerHealth> serverHealthMap =
-                (ConcurrentMap<String, ServerHealth>)
-                    packet.get(ProcessingConstants.SERVER_HEALTH_MAP);
-
-            serverHealthMap.put(
-                (String) packet.get(ProcessingConstants.SERVER_NAME), pair.getRight());
-            AtomicInteger remainingServersHealthToRead =
-                packet.getValue(ProcessingConstants.REMAINING_SERVERS_HEALTH_TO_READ);
-            remainingServersHealthToRead.getAndDecrement();
+            return doSuspend(
+                (fiber) -> {
+                  try {
+                    String url = serviceUrl + getRetrieveHealthSearchUrl();
+                    HttpRequest request =
+                        HttpRequest.newBuilder()
+                            .uri(URI.create(url))
+                            .timeout(Duration.ofSeconds(HTTP_TIMEOUT_SECONDS))
+                            .header("Authorization", "Basic " + encodedCredentials)
+                            .header("Accept", "application/json")
+                            .header("Content-Type", "application/json")
+                            .header("X-Requested-By", "WebLogic Operator")
+                            .POST(
+                                HttpRequest.BodyPublishers.ofString(
+                                    getRetrieveHealthSearchPayload()))
+                            .build();
+                    httpClient
+                        .sendAsync(request, HttpResponse.BodyHandlers.ofString())
+                        .thenApply(
+                            response ->
+                                new Result(
+                                    response.body(),
+                                    response.statusCode(),
+                                    response.statusCode() == ProcessingConstants.HTTP_OK))
+                        .whenComplete(
```

Comment:
Is there any support for a timeout?  What if the response never comes back?

Reply:
Also, I don't yet understand the flow if the result is not HTTP_OK.

Reply:
> Is there any support for a timeout? What if the response never comes back?
Timeout is set on request object and it's currently set to 1 min. I'll test the negative case to make sure request times out after specified duration and doesn't hang forever. Please let me know if current timeout value needs to be changed.

                        HttpRequest.newBuilder()
                            .uri(URI.create(url))
                            .timeout(Duration.ofSeconds(HTTP_TIMEOUT_SECONDS))

Reply:
Tested negative case where request times out due to HttpConnectTimeoutException. It logs below message in operator logs every 3 seconds.

{"timestamp":"05-13-2020T17:57:50.600+0000","thread":189,"fiber":"","domainUID":"","level":"SEVERE","class":"oracle.kubernetes.operator.steps.ReadHealthStep$ReadHealthWithHttpClientStep","method":"lambda$apply$1","timeInMillis":1589392670600,"message":"HTTP POST method on URL http://192.168.3.200:7001/management/weblogic/latest/serverRuntime/search failed with return code java.util.concurrent.CompletionException: java.net.http.HttpConnectTimeoutException: HTTP connect timed out.","exception":"","code":"","headers":{},"body":""}